crates/fuzzing/wasm-spec-interpreter/src/lib.rs

```diff
@@ -0,0 +1,85 @@
+//! This library provides a way to interpret Wasm functions in the official Wasm specificiation
+//! interpreter, written in OCaml, from Rust.
+
+use ocaml_interop::{OCamlRuntime, ToOCaml};
+
+/// Interpret the first function in the passed WebAssembly module (in Wasm form, currently, not
+/// WAT) with the given parameters.
+pub fn interpret(module: &[u8], parameters: Vec<Value>) -> Result<Vec<Value>, String> {
+    // Here we use an unsafe approach to initializing the `OCamlRuntime` based on the discussion in
+    // https://github.com/tezedge/ocaml-interop/issues/35. This was the recommendation to resolve
+    // seeing errors like `boxroot is not setup` followed by a `SIGSEGV`; this is similar to the
+    // testing approach in
+    // https://github.com/tezedge/ocaml-interop/blob/master/testing/rust-caller/src/lib.rs and is
+    // only as safe as the OCaml code running underneath; therefore, thhis approach is still unsafe
+    // and can cause SIGSEGV when running `cargo test` (TODO).
+    OCamlRuntime::init_persistent();
+    let ocaml_runtime = unsafe { OCamlRuntime::recover_handle() };
```

Comment:
This is an issue, but I'm not exactly sure how to resolve it. I originally used `OCamlRuntime::init` here but allocating and dropping the runtime in the same process is problematic. The approach above works (I haven't seen any crashes due to this in my local fuzzing) but `cargo test` occasionally crashes (I don't understand why).

Reply:
IIRC, the Ocaml runtime is not threadsafe (they're working on a version that is but it's a long-running project). Do we need to protect the runtime with a mutex?

Reply:
I tried `cargo test -j 1` to no avail as well as:

```
lazy_static! {
    static ref MUTEX: Mutex<u32> = Mutex::new(42);
}
```
with `let _ = MUTEX.lock().unwrap();` at the top of the `interpret` function. I still see failures when I test but not when I fuzz.

Reply:
Ok, this is resolved by the latest push (thanks @alexcrichton and @cfallin for the debugging help!):
 1. we _do_ need to maintain a mutex to avoid re-entrancy to the OCaml runtime but this needs to be locked with `let _lock = ...`, not `let _ = ...`
 2. we need to run Wasmtime last in the oracle so that its signal handlers do not get messed up by the handlers OCaml registers.